enhancements/network/ocp-on-smart-nics.md

```diff
@@ -0,0 +1,276 @@
+---
+title: Installing OCP on ARM-Based Smart NICs
+authors:
+  - "@danwinship"
+reviewers:
+  - "@russellb"
+  - "@dcbw"
+  - "@zshi-redhat"
+approvers:
+  - "@russellb"
+  - "@dcbw"
+creation-date: 2021-04-15
+last-updated: 2021-04-15
+status: provisional
+---
+
+# Installing OCP on ARM-Based Smart NICs
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As servers have greater and greater networking needs, hardware vendors
+have created increasingly more complex and powerful network cards. The
+latest iteration of this is "Smart NICs" that include an entire ARM
+system running Linux on the NIC. With such a NIC, we can offload
+nearly all network processing from the host computer onto the NIC
+"DPU" (Data Processing Unit), freeing up host CPU resources for
+customer workloads (and allowing faster data transfer by ensuring the
+network stack doesn't need to compete for CPU with the customer
+workload).
+
+However, doing this requires installing and managing an appropriate
+software stack on all of the NICs in the cluster. This software will
+need to be kept in sync with the software on the host PC (eg, as new
+functionality is added to OCP). The obvious answer to this problem is
+to run OpenShift on the NIC as well as on the host.
```

Comment:
> This answer is not obvious to me.

Yeah, the more I wrote the less obvious it became. :slightly_smiling_face:

This decision had already been made before I got involved, and I don't yet know everything that has been discussed, and I'm discovering now that this has apparently already been discussed more with other OpenShift teams that I hadn't even heard about. So this WIP enhancement is partly about trying to gather all this info...

> Will we have the same remote-worker-node issue that resulted in creating single-node deployments?

Not sure what that refers to

> Do we need to support these things being part of multi-node as well as single-node deployments?

I'm not sure from your questions if you're talking about the "host" cluster being multi/single-node or the NIC cluster. The expected architecture for the host cluster is that you'll have 3 masters, some non-0 number of dumb workers, and a medium to large number of smart workers with smart NICs. The NICs would either all be part of the same OCP cluster as the hosts, or else would all be part of some other OCP cluster. I don't think there's any interest in running a single-node cluster on a NIC. I'll clarify that.

> Will these NICs have the RAM and CPU capacity to run as an OpenShift node?

Yes. (BF-2 has "Up to 8 Armv8 A72 cores" and "8GB / 16GB / 32GB of on-board DDR4", and we could say we only supported a subset of configurations if we needed to.) I should make this more explicit in the "Goals"/"Non-Goals".

> Do we know for certain that we will be able to "own" the OS on the NIC? For all vendors? Will all of the drivers we need be open sourced?

For BF-2, yes. For other vendors, not necessarily. The "Goals"/"Non-Goals" section is supposed to be trying to carve out the set of NICs we might support. (eg, there's a Non-Goal of "`Support for ARM-based Smart NICs where we would not be able to run RHCOS on the NIC (eg, because the NIC requires proprietary drivers).`")

> Do we need to support running software on the NICs that is not containerized?

No more so than on any other OCP node. (eg, we'd run OVS as a systemd service, but ovn-kubernetes in pods.) In particular, the ARM OS doesn't need to run any support software for the NIC.

Reply:
> > This answer is not obvious to me.
> 
> Yeah, the more I wrote the less obvious it became. 🙂
> 
> This decision had already been made before I got involved, and I don't yet know everything that has been discussed, and I'm discovering now that this has apparently already been discussed more with other OpenShift teams that I hadn't even heard about. So this WIP enhancement is partly about trying to gather all this info...

+1 to collecting it all in one place

> > Will we have the same remote-worker-node issue that resulted in creating single-node deployments?
> 
> Not sure what that refers to

If a node reboots and can't connect to its control plane, it doesn't start any workloads. Since we can't ensure that all RAN nodes would always be able to see their control plane, we couldn't ensure that they would bring their workloads back online after a power loss. That resulted in [a series of enhancements](https://github.com/openshift/enhancements/tree/master/enhancements/single-node) to run the entire OpenShift stack on a single node.

> > Do we need to support these things being part of multi-node as well as single-node deployments?
> 
> I'm not sure from your questions if you're talking about the "host" cluster being multi/single-node or the NIC cluster. The expected architecture for the host cluster is that you'll have 3 masters, some non-0 number of dumb workers, and a medium to large number of smart workers with smart NICs. The NICs would either all be part of the same OCP cluster as the hosts, or else would all be part of some other OCP cluster. I don't think there's any interest in running a single-node cluster on a NIC. I'll clarify that.

It seems like these NICs could have the same connectivity issues, especially if they are used in one of the RAN servers on a cell tower. As you say, the configurations described in the enhancement assume the NIC is either a member of the same cluster its host is or the member of a different multi-node cluster. The first approach doesn't seem likely to work for the telco cases at all because it would result in a 2 node cluster, with all of the reliability issues that brings with it. The second approach would only work if the NIC is running its own single-node deployment.

If our telco customers are not looking at these NICs, then maybe that limitation doesn't matter, for now.

> > Will these NICs have the RAM and CPU capacity to run as an OpenShift node?
> 
> Yes. (BF-2 has "Up to 8 Armv8 A72 cores" and "8GB / 16GB / 32GB of on-board DDR4", and we could say we only supported a subset of configurations if we needed to.) I should make this more explicit in the "Goals"/"Non-Goals".

Nice. Can I get one to use as a development server? ;-)

Reply:
> The first approach doesn't seem likely to work for the telco cases at all because it would result in a 2 node cluster,

Maybe, in case of a SNO,  we can have it as a worker (host - master, card - worker). It's not a real SNO of course, but a master + worker topology (that is doesnt yet exist of course).